gplugins/lumerical/fdtd.py

```diff
@@ -0,0 +1,1282 @@
+                zmin=0.0,
+                zmin_tolerance=None,
+                material="si",
+                sidewall_angle=10.0,
+                sidewall_angle_tolerance=None,
+                width_to_z=0.5,
+                z_to_bias=None,
+                mesh_order=2,
+                layer_type="grow",
+                mode=None,
+                into=None,
+                resistivity=None,
+                bias=None,
+                derived_layer=None,
+                info={"active": True},
+                background_doping_concentration=100000000000000.0,
+                background_doping_ion="Boron",
+                orientation="100",
+            ),
+            # KNOWN ISSUE: Lumerical 2021 version of Layer Builder does not support dopants in process file
+        }
+    )
+    SIMULATION_SETTINGS_LUMERICAL_FDTD.port_translation = 1.0
+    sim = LumericalFdtdSimulation(
+        component=taper,
+        layerstack=layerstack_lumerical2021,
+        convergence_settings=LUMERICAL_FDTD_CONVERGENCE_SETTINGS,
+        simulation_settings=SIMULATION_SETTINGS_LUMERICAL_FDTD,
+        hide=False,
+        run_port_convergence=False,
+        run_mesh_convergence=False,
+    )
+
+    sp = sim.write_sparameters(overwrite=True)
+    print(sp)
+    print("Done")
+
+
+class LumericalFdtdSimulation:
+    """
+    Lumerical FDTD simulation
+
+    Set up FDTD simulation based on component geometry and simulation settings. Optionally, run convergence.
+
+    Attributes:
+        component: Component geometry to simulate
+        layerstack: PDK layerstack
+        session: Lumerical session
+        simulation_settings: FDTD simulation settings
+        convergence_settings: FDTD convergence settings
+        dirpath: Directory where simulation files are saved
+        filepath_npz: S-parameter filepath (npz)
+        filepath_fsp: FDTD simulation filepath (fsp)
+        mesh_convergence_data: Mesh convergence results
+
+    """
+
+    def __init__(
+        self,
+        component: Component,
+        layerstack: LayerStack | None = None,
+        session: lumapi.FDTD | None = None,
+        simulation_settings: SimulationSettingsLumericalFdtd = SIMULATION_SETTINGS_LUMERICAL_FDTD,
+        convergence_settings: ConvergenceSettingsLumericalFdtd = LUMERICAL_FDTD_CONVERGENCE_SETTINGS,
+        dirpath: PathType | None = "",
+        hide: bool = True,
+        run_mesh_convergence: bool = False,
+        run_port_convergence: bool = False,
+        run_field_intensity_convergence: bool = False,
+        xmargin: float = 0,
+        ymargin: float = 0,
+        xmargin_left: float = 0,
+        xmargin_right: float = 0,
+        ymargin_top: float = 0,
+        ymargin_bot: float = 0,
+        zmargin: float = 1.0,
+        **settings,
+    ):
+        r"""Creates FDTD simulation for extracting s-parameters
+
+        Your components need to have ports, that will extend over the PML.
+
+        .. image:: https://i.imgur.com/dHAzZRw.png
```

Comment:
what does this show? the ports extending beyond PML?

Reply:
does this lead to an autogenerated doc or something that's more clear about what this shows?

Reply:
Yes, I see this in my PyCharm IDE. Other IDEs may compile this image similarly.